### DISCUSSION TITLE
New syntax for immutable arrays

### DISCUSSION THREAD

**Discussion by TahirAhmadov** [dotnet/csharplang#5451] [Language Ideas]
Can we get a symbol to create immutable arrays without the type name?
```cs
static readonly int[]~ a = new[]~ { 1, 2, 3, }; 
// same as
static readonly ImmutableArray<int> a = ImmutableArray.Create(new[] { 1, 2, 3 });
// notice how much more boiler plate there is to express the same idea?
```
I don't have any preference for the symbol. 
The benefit would be allowing the compiler to optimize this variable: it can be lifted into program data, instead of runtime heap initialization. Also, it obviously saves a lot of typing over using the type explicitly.

**Comment by CyrusNajmabadi** [comment]
Proposal for this is already in progress here: https://github.com/dotnet/csharplang/issues/5354

**Reply by TahirAhmadov** [reply]
Hmm, it isn't? That talks about collection literals which can be target typed to an `ImmutableArray<>`, among other things, but it doesn't propose creating a new syntax for declaring `ImmutableArray<>` variables.
```cs
ImmutableArray<int> a = [1,2,3]; // issue #5354 proposes this
int[]~ a = new []~ { 1, 2, 3 }; // this discussion proposes this
int[]~ a = [1, 2, 3]; // this becomes possible if both the proposals in this discussion and in #5354 are done, which would be awesome
```

**Reply by 333fred** [reply]
I guess. But I would also say that we are not interested in omitting the type name.

**Reply by CyrusNajmabadi** [reply]
> // this discussion proposes this

That doesn't seem much better :)

**Reply by TahirAhmadov** [reply]
> That doesn't seem much better :)

It's not better or worse; it complements the collection literals.

**Comment by svick** [comment]
> The benefit would be allowing the compiler to optimize this variable: it can be lifted into program data, instead of runtime heap initialization. 

I don't think that would be possible, since `ImmutableArray<T>` actually needs the managed array to exist.

It is possible to optimize this code using program data, but that optimization is already implemented: code like `ImmutableArray.Create(new[] { 1, 2, 3 })` or `ImmutableArray.Create(1, 2, 3, 4, 5)` [is already compiled using `RuntimeHelpers.InitializeArray`](https://sharplab.io/#v2:D4AQDABCCMB0DCB7ANsgpgYwC4EtEDsBnWASQFsyBXLAQwCN0BuAWACg3yraG0BBAJ340Angn5oaWNAAp8aAO4BtALoQA3hGgAaCACYdAZggBfAJQtWnavXQCho+OMkztewzoAsOgKzmgA==).

> Also, it obviously saves a lot of typing over using the type explicitly.

Yes, but it also makes the code much harder to understand. I don't think that trade-off is worth it here.

**Reply by TahirAhmadov** [reply]
> Yes, but it also makes the code much harder to understand. I don't think that trade-off is worth it here.

I don't see how that's the case. There is no convertibility between the 2 types, so accidental assignments are impossible. The symbol (whichever one is chosen) clearly indicates it's an immutable array. 
Think about the upsides. Countless numbers of "immutable by intent" array properties (both get-only and get&set ones) can be converted easily to immutable arrays by sprinkling this symbol; we don't have to write `ImmutableArray<>` or `IReadOnlyList<>` anymore.

**Reply by HaloFour** [reply]
I'd be much more concerned about the human parser than the compiler.  It'd be very easy to scan `int[]~` and misinterpret it.  I'd rather the intent be clear, which it is with `ImmutableArray<T>` or `IReadOnlyList<T>`.

**Reply by TahirAhmadov** [reply]
> I'd be much more concerned about the human parser than the compiler.

The same can be said about any symbol. Obviously we don't want a language where everything is an English word. Immutable arrays are important enough to warrant their own symbol, not just an identifier.

**Reply by HaloFour** [reply]
It's not the fact that it's a symbol, it's the fact that it's a collection of symbols which looks _very_ similar to an existing collection of symbols that means something different.  Many things in C# are keywords specifically because it is easier to read and understand at a glance.

**Reply by TahirAhmadov** [reply]
> it's the fact that it's a collection of symbols which looks very similar to an existing collection of symbols that means something different.

That may be a valid concern. Like I said, I'm not stuck on `[]~` or any particular symbols. If you can think of a symbol combination which you think is more suitable from readability perspective, I'm all for it. My point is, these identifiers are way too clunky, too the point of being a detriment to adoption.

**Reply by CyrusNajmabadi** [reply]
>  point is, these identifiers are way too clunky, too the point of being a detriment to adoption.

That seems highly unlikely. List is much more used that primitive arrays, despite the latter having cute syntax for them.  Users have no issues using framework types, despite having to use their names.

**Reply by TahirAhmadov** [reply]
> That seems highly unlikely. List is much more used that primitive arrays, despite the latter having cute syntax for them. Users have no issues using framework types, despite having to use their names.

Two important differences. `List<>` provides very important benefits over arrays: it's mutable (for when that's needed); LINQ `ToList` is more efficient because it doesn't need to double allocate like `ToArray`. Also, `List<>` is much shorter than `ImmutableArray<>`. On the other hand, immutable arrays do not serve a purpose which is directly needed, like mutability; everything it does, the regular array also does - provided you don't need to add/remove items. So the only reason an advanced developer would use an immutable array over a regular one is perfecting code quality; that is a much weaker reason than those for using `List<>` (when it's needed). Given the weaker reasons, the barrier to entry should be lower, too - like a simple symbol-based syntax to declare.

**Reply by CyrusNajmabadi** [reply]
Barrier to entry is not lowered with specialized syntax. It's an additional thing to learn as being special. I still don't get the train that someone would have trouble learning and using this, but is fine with List

**Reply by TahirAhmadov** [reply]
I think "barrier to entry" was a bad term to describe what I meant. More appropriate word would be "price". A developer knows he wants immutable arrays; the question is, how much is he willing to pay for it. Given that the value is reducing potential bugs caused by third parties modifying arrays that are not meant to be modified, it's a "low return". Especially so compared with `List<>` which gives you important features; it's a "high return". Counterintuitively, the price for getting `List` is 4 letters, whereas for immutable array, it's 14 letters. So, you pay little and get a lot with `List`, and pay a lot but get a little with `ImmutableArray`.

**Reply by theunrepentantgeek** [reply]
The value of immutable data structures is much much higher than you've described. 

There's a moderately well known case study from Facebook describing a C++ codebase where a switch to immutable data structures cut memory use (increased sharing, especially across threads), improved performance (largely by eliminating useless data copying, but also by improving cache locality) and reduced bugs (mostly through better computer enforcement). 

Admittedly, these results are not typical. 😁 

Even with all that, however, I don't think dedicated syntax for just immutable array literals is justifiable. The general purpose proposal for collection literals will be more widely useful.

**Reply by TahirAhmadov** [reply]
> The value of immutable data structures is much much higher than you've described.

I don't dispute that at all. Experienced developers know that the value of immutability is high. My point is that, on an every day basis, given the time and resource constraints, some corners will be cut - that's a given. On the list of priorities of a typical developer in the real world, the perceived value of immutable arrays is lower than that of `List<>` because of the reasons I specified; if he needs to add items to a collection, he **has** to use a `List<>`, but there is nothing which forces him to use `ImmutableArray<>` - other than his desire to reduce potential problems down the line. 

> There's a moderately well known case study from Facebook describing a C++ codebase where a switch to immutable data structures cut memory use (increased sharing, especially across threads), improved performance (largely by eliminating useless data copying, but also by improving cache locality) and reduced bugs (mostly through better computer enforcement).

Can you please link to it? In any case, this is a strong argument for making `ImmutableArray<>` a first class citizen in C#.

> Even wit h slob that, however, I don't think dedicated syntax for just immutable array literals is justifiable. The general purpose proposal for collection literals will be more widely useful.

I addressed this previously, but it bears repeating: this proposal does not conflict with collection literals #5354 in any way, and in fact works together with it beautifully:
```cs
ImmutableArray<int> a = [1,2,3]; // issue #5354 proposes this
int[]~ a = new []~ { 1, 2, 3 }; // this discussion proposes this
int[]~ a = [1, 2, 3]; // this becomes possible if both the proposals in this discussion and in #5354 are done, which would be awesome
```

**Reply by CyrusNajmabadi** [reply]
I think you're overstating the problem using ImmutableArray.  Nearly everything a programmer uses comes from teh BCL or types from libs.  THe bar to entry in the language is enormously high and legit has to feel like a lib solution is not suitable.  I don't see that here. ImmutbaleArray is trivial to use.  If it's the right fit for a developer, they can just reference it (with intellisense making it trivial to write as well).

**Reply by najak3d** [reply]
I'd vote for this proposal.   I'm annoyed that the most useful of classes are stored into deep namespaces, like:

System.Collections.Generic
and
System.Collections.Immutable

It's clear these namespaces were an "after thought".    If we could do it all again, we'd swap stuff up... and put all the good stuff into:

System.Collections

(rather than all the bad original stuff that nobody uses now, like non-Generic "List")

So if a C# spec can be created to help overcome this -- I'll vote for it.

**Reply by CyrusNajmabadi** [reply]
@najak3d proposal is here: https://github.com/dotnet/csharplang/issues/5354
